File: proposed-projects/round-trip-experiment/reverse/20260909T124157Z_4ee67ba2/20260910T071826Z_7d46923a/ported_implementation/gbnf/gbnf.py
from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder import GrammarParseError, RulesBuilder


def GBNF(grammar: str, initial_string: str = "") -> ParseState:
    if not isinstance(grammar, str):
        raise Exception("grammar must be a string")

    if not isinstance(initial_string, str):
        raise Exception("input must be a string")

    rules_builder = RulesBuilder(grammar)
    rules = rules_builder.rules
    symbol_ids = rules_builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, "No rules were found")
    root_id = symbol_ids.get("root")
    if root_id is None:
        raise GrammarParseError(
            grammar, 0, "Grammar does not contain a 'root' symbol"
        )

    stacked_rules = [build_rule_stack(rule) for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))

File: packages/gbnf-experiment/data/20260909T130706Z_31ac2789/ported_implementation/src/gbnf.ts
import { Graph } from './grammar_graph/graph.ts';
import { ParseState } from './grammar_graph/parse_state.ts';
import { buildRuleStack } from './grammar_parser/build_rule_stack.ts';
import { GrammarParseError, RulesBuilder } from './rules_builder/index.ts';

export const GBNF = (grammar: string, initialString = ''): ParseState => {
  if (typeof grammar !== 'string') {
    throw new Error('grammar must be a string');
  }

  if (typeof initialString !== 'string') {
    throw new Error('input must be a string');
  }

  const rulesBuilder = new RulesBuilder(grammar);
  const { rules, symbolIds } = rulesBuilder;
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  if (symbolIds.get('root') === undefined) {
    throw new GrammarParseError(grammar, 0, "Grammar does not contain a 'root' symbol");
  }
  const rootId: number = symbolIds.get('root') as number;

  const stackedRules = rules.map((rule) => buildRuleStack(rule));
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

export default GBNF;
